docs: Created New Documentation for RDMA (#517)
openebs/website: docs/main/user-guides/replicated-storage-user-guide/replicated-pv-mayastor/configuration/rs-rdma.md

---
id: rs-rdma
title: Enable RDMA for Volume Targets
keywords:
 - OpenEBS Replicated PV Mayastor
 - Replicated PV Mayastor Configuration
 - Configuration
 - RDMA
 - Enable RDMA
 - Enable RDMA for OpenEBS Replicated PV Mayastor
description: This guide will help you to enable RDMA for OpenEBS Replicated PV Mayastor.
---

# Enable RDMA for Volume Targets

> The RDMA feature in Replicated PV Mayastor enables you to achieve better throughput and latency outcomes for the applications using Replicated PV Mayastor Volumes. This feature provides a significant performance boost by leveraging RDMA technology. A kubernetes cluster consisting of nodes that have RNIC (RDMA Network Interface Card) can make use of this feature.

Reviewer: Always upper-case K in Kubernetes

> :::info

Reviewer 2: Please remove this note from here.

Before enabling and using RDMA in Replicated PV Mayastor, ensure the following prerequisites are met:

1. **Interface Validation**

  Ensure that the interface name specified in `io-engine.target.nvmf.iface` exists on all io-engine nodes. Misconfiguration of this parameter can lead to unexpected behavior.

>   If the interface name is invalid or not RDMA-capable on a node, Replicated PV Mayastor will run with only default tcp support for such nodes.

Reviewer: "will run with only default TCP support" means what?

Author: Modified this sentence as **"If the interface name is invalid or not RDMA-capable on a node, Replicated PV Mayastor volume targets on such nodes will support only TCP connections."**

Reviewer: Not sure about invalid interface name, that can result in bigger issues.
Just stick to 'not RDMA-capable'.

>   If the interface name is invalid or not RDMA-capable on a node, Replicated PV Mayastor will run with only default tcp support for such nodes.

Reviewer 2: ```suggestion
  If the interface name is invalid or not RDMA-capable on a node, Replicated PV Mayastor volume targets on such nodes will support only tcp connections.
```

2. **Application Node Requirements**

  Application nodes must also be RDMA-capable to connect to RDMA-enabled volume targets. This capability is independent of the Helm-provided interface name for `io-engine.target.nvmf.iface`. RDMA connections will work as long as the application node has an RDMA-capable device.

3. **TCP Fallback Behavior**

>   If an application is hosted on a non-RDMA-capable node, connections to RDMA-enabled volume targets will default to TCP. To prevent this fallback behavior, set the Helm chart value `csi.node.nvme.tcpFallback` to `false`. In this configuration, the initiator (application) node will continuously attempt RDMA connections, resulting in connection failures. To mitigate this, you can either keep the fallback option enabled or relocate the application pod to an RDMA-capable node, provided the pod's resource type allows for such movement.

Reviewer: Should this "_To prevent this fallback behavior, set the Helm chart value `csi.node.nvme.tcpFallback` to `false`_" be rephrased as " The TCP fallback behaviour can be disabled by..." ?

4. **Soft-RoCEv2 Support**

>   RDMA devices can be manually created using `ibverbs` utilities. For example, to create an RDMA device named `rxe0` on an Ethernet interface `eth0`, run:

Reviewer: Need to explain the need for this. Something on the lines "Mayastor also supports software RDMA if the node does not have RDMA-capable hardware network interface cards."

4. **Soft-RoCEv2 Support**

  RDMA devices can be manually created using `ibverbs` utilities. For example, to create an RDMA device named `rxe0` on an Ethernet interface `eth0`, run:

  ```
  rdma link add rxe0 type rxe netdev eth0
  ```

  Once created, `rxe0` can be utilized as outlined in the Enablement section.
  
  :::note
  The IP address assignments to RDMA GIDs on these devices depends on the cluster's networking and CNI configurations. Variations in Soft-RoCEv2 devices are not fully understood and have not been extensively tested.
  :::

## Enable RDMA

To enable the RDMA feature via Helm:

1. Set `io-engine.target.nvmf.rdma.enabled` to `true`.

2. Set `io-engine.target.nvmf.iface` to a valid network interface name that exists on an RNIC.

3. Verify that all nodes are properly configured with RDMA-capable hardware and that network interfaces are correctly identified and accessible.

> 

Reviewer 2: and move that note here, without mentioning `Default Behaviour`

```
Once enabled, all Mayastor volumes will connect over RDMA.
:::
```

> ## Performance Indications

Reviewer: What does Indications mean?

Reviewer 2: @balaharish7 You can simply write `Performance Results`
